---
title: "TDP-43 binding in 293"
output: html_notebook
author: Giulia Manferrari
date: 21.07.21
---

# CLIP META-ANALYISIS

## DATA 

Public available CLIP data targeting TDP-43 from the following studies: 
Tollervey et al 2017: link to study / database code 
Rot et al 2019


### Libraries
```{r LIBRARY, message=FALSE, warning=FALSE}
library(dplyr)    
library(ggplot2) 
library(DESeq2)
library(data.table)
library(dplyr)
library(tidyverse)
library(tidyr)
library(reshape2)
library(BiocManager)
library(ggpubr)
library(ggsci)
library(RColorBrewer)
library(plotly)
library(ggplot2)
library(tximport)
library(GenomicRanges)
library(rtracklayer)
library(GenomicFeatures)
library(readr)
library(scales) 
library(cowplot)
```

### Working directory 
```{r}
setwd("/Users/manferg/clip_metanalysis/tollervey_brain/")
```

### Functions
```{r}
reorder_within <- function(x, by, within, fun = mean, sep = "___", ...) {
  new_x <- paste(x, within, sep = sep)
  stats::reorder(new_x, by, FUN = fun)
}

scale_x_reordered <- function(..., sep = "___") {
  reg <- paste0(sep, ".+$")
  ggplot2::scale_x_discrete(labels = function(x) gsub(reg, "", x), ...)
}
```
### Colour palettes
```{r}
mypal.dedup<-c("input_reads"= "#5FB1D4",  "output_reads"= "#39373B") 
mypal.dedup.ratio<-c("ratio"= "#5FB1D4")
my.pal.lib<-c("output_reads"= "#39373B")
mypal.region<-c("intron" = "#454241","CDS" ="#F0421C", "intergenic" ="#DDD3D1", "ncRNA" = "#3DB75D", "UTR5" = "#3DB7E6", "UTR3"= "#D644C7")
mypal.model<-c("FTLD-TDP_human_brain" = "#F21905","Healthy_human_brain" = "#49BF45","SH-SY5Y"="#9263A6","293Flp" ="#034C8C", "hES" ="#04ADBF")

mypal.biotype<-c("lncRNA"="#BF4215","mRNA" = "#224A59","pre-mRNA"= "#80b1d3",
                "miRNA" ="#fb8072", "mt_rRNA" =  "#9DA65D", "mt_tRNA" = "#ccebc5",
                "snRNA" = "#D99543","snoRNA"="#BF812D" ,"NA" = "#FFFEFF", "sRNA"="#d9d9d9",    "rRNA"="#bc80bd", "translated_unprocessed_pseudogene"  ="#ffed6f")

```

### ordering index 
```{r}

reorder_sample_idx <- c("tollervey_brain1","tollervey_brain2","tollervey_brain6.high","tollervey_brain7.low","tollervey_brain3","tollervey_brain4","tollervey_brain5")

```


# IMPORT INTERSECTED BED FILES

```{r}
intresected.chr.df<-read_csv("/Users/manferg/clip_metanalysis/intresected.chr.df.csv")

intresected.chr.df<-intresected.chr.df %>% as.data.frame() %>% dplyr::select(-X1)

unique(intresected.chr.df$biotype)

#rename tollervey_SHSY5Y1a and tollervey_SHSY5Y1b to high and low.
intresected.chr.df<-intresected.chr.df %>% mutate(sample = gsub("tollervey_SHSY5Y1a", "tollervey_SHSY5Y1a.high",sample), sample = gsub("tollervey_SHSY5Y1b", "tollervey_SHSY5Y1b.low", sample))

dim(intresected.chr.df) #4465898
```

### IMPORT METADATA
```{r}
#METADATA----------------------
# library(readxl)
# df = read_excel("/Users/manferg/clip_metanalysis/meta_metadata1.xlsx", sheet=1)
#   write.csv(df, gsub("xlsx", "csv", "/Users/manferg/clip_metanalysis/meta_metadata.csv"), row.names=FALSE)

metadata<- read_csv("/Users/manferg/clip_metanalysis/meta_metadata.csv") %>% as.data.frame %>% column_to_rownames(var="meta_id")
rownames(metadata)
colnames(metadata)

# metadata<- metadata[reorder_sample_idx ,]
metadata$meta_id <- rownames(metadata)
metadata<- metadata %>% mutate(meta_id = gsub("tollervey_SHSY5Y1a", "tollervey_SHSY5Y1a.high",meta_id), meta_id = gsub("tollervey_SHSY5Y1b", "tollervey_SHSY5Y1b.low", meta_id))
metadata$meta_id -> rownames(metadata)

```

### Merging BED INTERSECT and METADATA 

```{r}

intresected.chr.df<-left_join(intresected.chr.df, metadata, by=c("sample" = "meta_id"))

intresected.chr.df.fil<-dplyr::select(intresected.chr.df,-species,-technology, -study_id, -barcode)#filter metada columns to exclude

unique(intresected.chr.df.fil$sample)

dim(intresected.chr.df) #4465898


```
### SELECT SAMPLES 

```{r}
intresected.chr.df.fil<-intresected.chr.df.fil %>% filter( .,grepl("^grot_293fl",sample))
unique(intresected.chr.df.fil$sample)

```

### DATASET LISTS 

Split each samples dataset as an element of a list 

```{r}
#main df list----------------

main.chr.li = split(intresected.chr.df.fil,intresected.chr.df.fil$sample) #transform df into list 


# main.chr.li<-main.chr.li[reorder_sample_idx] #reorder list elements 
# names(main.chr.li) #check order

```

# number of xlinks per gene 
```{r}
#counts (or number of xlinks) for each gene/region------------------CHR
xlink.events.gene.chr.li =list()
top.xlink.events.gene.chr.li=list()
for (i in 1:length(main.chr.li)){
  xlink.events.gene.chr.li[[i]]<-main.chr.li[[i]] %>% group_by(gene_name) %>% summarize(n=n()) %>% arrange( .,desc(n)) %>% as.data.frame()#counts (or number of xlinks) for each gene.. 
  sample<-as.character(names(main.chr.li[i])) # Create a new vector with sample names
  xlink.events.gene.chr.li[[i]]$sample <- sample
  top.xlink.events.gene.chr.li[[i]]<-xlink.events.gene.chr.li[[i]][2:10,]  #exclude "None" and list top 20 gene..s
  top.xlink.events.gene.chr.li[[i]]$gene_name <- factor(top.xlink.events.gene.chr.li[[i]]$gene_name , levels = top.xlink.events.gene.chr.li[[i]]$gene_name[order(top.xlink.events.gene.chr.li[[i]]$n)])

}

sapply(xlink.events.gene.chr.li, nrow)#different number of genes in each datasets
xlink.events.chr.df<-do.call(rbind,xlink.events.gene.chr.li) #convert back to df to plot



#reorder levels as samples order
xlink.events.chr.df$sample <- factor(xlink.events.chr.df$sample , levels=unique(xlink.events.chr.df$sample))
xlink.events.chr.df$gene_name <- factor(xlink.events.chr.df$gene_name , levels=unique(xlink.events.chr.df$gene_name))

top.xlink.events.chr.df<-do.call(rbind,top.xlink.events.gene.chr.li)
top.xlink.events.chr.df<-left_join(top.xlink.events.chr.df,metadata,by=c("sample" = "meta_id"))
#reorder levels as samples order
top.xlink.events.chr.df$sample <- factor(top.xlink.events.chr.df$sample , levels=unique(top.xlink.events.chr.df$sample ))

#data exploration xlink df----------

dim(xlink.events.chr.df)  #72170 

length(unique(xlink.events.chr.df$gene_name)) #20992 unique gene_name


```

# score per gene 

```{r}

score.events.gene.chr.li =list()
top.score.events.gene.chr.li=list()
for (i in 1:length(main.chr.li)){
  score.events.gene.chr.li[[i]]<-main.chr.li[[i]] %>% group_by(gene_name) %>% summarize(score=sum(score)) %>% arrange( .,desc(score)) %>% as.data.frame()#counts (or number of scores) for each gene.. 
  sample<-as.character(names(main.chr.li[i])) # Create a new vector with sample names
  score.events.gene.chr.li[[i]]$sample <- sample
}

sapply(score.events.gene.chr.li, nrow)#different number of genes in each datasets
score.events.chr.df<-do.call(rbind,score.events.gene.chr.li) #convert back to df to plot

#reorder levels as samples order
score.events.chr.df$sample <- factor(score.events.chr.df$sample , levels=unique(score.events.chr.df$sample ))
score.events.chr.df$gene_name <- factor(score.events.chr.df$gene_name , levels=unique(score.events.chr.df$gene_name ))
length(score.events.chr.df$gene_name) #72170
dim(score.events.chr.df) #72170 

```

# Merging score and xlink per gene datasets 
```{r}
score.xlink.df<-inner_join(xlink.events.chr.df,score.events.chr.df , by = c("sample" = "sample","gene_name" = "gene_name")) %>% na.omit() #remove intergenic region to plot frequencies 


dim(score.xlink.df) #72170
dim(xlink.events.chr.df)  #72170
dim(score.events.chr.df) #72170 



score.xlink.df[score.xlink.df$gene_name =="MALAT1",]
score.xlink.df[score.xlink.df$gene_name =="SNORA54",]
score.xlink.df[score.xlink.df$gene_name =="AL356488.2",]
score.xlink.df[score.xlink.df$gene_name =="TAF1D",]
score.xlink.df[score.xlink.df$gene_name =="RBFOX1",]
```


#FILTERING OUT LOW CROSS-LINKS 

```{r}

# score.xlink.df.filt<-score.xlink.df %>% filter(score > 5) %>% filter(n > 5)
score.xlink.df.filt<-score.xlink.df %>% filter(score > 10) %>% filter(n > 10)
score.xlink.df.filt$n<-as.double(score.xlink.df.filt$n)

score.xlink.df.filt<- score.xlink.df.filt[score.xlink.df.filt$n !=score.xlink.df.filt$score, ]


score.xlink.df.filt<-arrange(score.xlink.df.filt, score.xlink.df.filt$score)
score.xlink.df.filt<-score.xlink.df.filt %>% filter(gene_name != "None") %>% arrange( .,desc(score))

dim(score.xlink.df) #55495 
dim(score.xlink.df.filt) #24072 



#merge score by sum


score.xlink.df.filt$sample<-gsub("grot_293fl_2",'grot_293fl_1',score.xlink.df.filt$sample)
score.xlink.merged.filt<-aggregate(score ~ sample + gene_name,FUN = "sum", data=score.xlink.df.filt) %>% arrange( .,desc(score)) #merging ftld samples 

dim(score.xlink.merged.filt)
tail(score.xlink.merged.filt)

write.csv(score.xlink.merged.filt, file="tdp43_targets_293.csv")
```

```{r}

dens_plot_xl<-ggplot(score.xlink.df,aes(x=log2(n+1))) +
  geom_density() +
  ggtitle("Cross-link sites frequency")+
  facet_wrap(~sample, scale="fixed") +
  xlab("Log2 xlinks") +
  ylab("Frequency") +
  theme_bw()
dens_plot_xl


dens_plot_xl_fil<-ggplot(score.xlink.df.filt,aes(x=log2(n+1))) +
  geom_density() +
  facet_wrap(~sample, scale="fixed") +
  ggtitle("Cross-link sites frequency post filtering")+
  xlab("Log2 xlinks") +
  ylab("Frequency") +
  theme_bw()
dens_plot_xl_fil


```


# SCORE FREQUENCY 
```{r}
threshold.10= 3.32

dens_plot_score<-ggplot(score.xlink.df,aes(x=log2(score))) +
  geom_density() +
  facet_wrap(~sample,scale="fixed") +
  xlab("Log2 score") +
  scale_x_continuous(breaks = scales::pretty_breaks(n = 10))+
  ylab("Frequency") +
  geom_vline(xintercept = threshold.10, linetype = "dashed", color ="red") +
  ggtitle("Score Frequency distribution") +
  theme_bw()+
  theme(plot.title = element_text(size=16))
  
dens_plot_score

dens_plot_score_fil<-ggplot(score.xlink.df.filt,aes(x=log2(score))) +
  geom_density() +
  facet_wrap(~sample,scale="fixed") +
  xlab("Log2 score") +
  scale_x_continuous(breaks = scales::pretty_breaks(n = 10))+
  ylab("Frequency") +
  geom_vline(xintercept = threshold.10, linetype = "dashed", color ="red") +
  ggtitle("Score Frequency distribution") +
  theme_bw()+
  theme(plot.title = element_text(size=16))
  
dens_plot_score_fil


```

```{r}
library(scattermore)

 scatter<- ggplot(score.xlink.df) +
  geom_scattermore(aes(x=log2(n),y=log2(score)), stat="identity",pointsize=1,colour="black") + facet_wrap(~sample) + theme_bw()
  
scatter


 scatter_fil<- ggplot(score.xlink.df.filt) +
  geom_scattermore(aes(x=log2(n),y=log2(score)), stat="identity",pointsize=1,colour="black") + facet_wrap(~sample) + theme_bw()
  
 scatter_fil
```


